### PR TITLE
Initial support for ROCm: rename CUDA to GPU et al, pimpl for GPU details

### PR DESCRIPTION
**Pull request by jeffdaily**
- add ROCm and ROCm TensorFlow support to setup.py
- Renamed symbols and files
  - CUDA -> GPU
  - Cuda -> Gpu
  - cuda -> gpu
  - HAVE_CUDA -> HAVE_GPU --- iff generically referring to any GPU device
- CUDAContext becomes GPUContext and uses pimpl pattern to decouple
  CUDA and HIP implementations of GPU interface.
- New GPUContext methods
  - StreamCreate
  - StreamSynchronize
  - GetDevice, SetDevice
  - MemcpyAsyncD2D, MemcpyAsyncH2D, MemcpyAsyncD2H

Signed-off-by: Jeff Daily <jeff.daily@amd.com>

### CODE REVIEW

**Comment by jeffdaily** [comment]
It's not clear to me whether the current test failures are related to the changes in this PR.

**Comment by tgaddair** [comment]
Hey @jeffdaily, thanks for the new PR.  Overall, I like the approach to abstracting GPU operations in this way.

One thing I noticed is it looks like the DDL codepaths haven't been updated to use the new `gpu_operations.h`.  Those aren't currently under unit test, but the changes should be fairly straightforward.

There was also a PR landed yesterday that added NCCL broadcast support which will require a rebase.

For the unit tests, it looks like at least one test was failing because the Horovod MXNet plugin failed to compile, so we may need to dig into why that is.  I'm rerunning some of the other tests to see if it was a transient error.

I'll try and put together inline feedback as well.  I think it would be good to have @nvcastet and @romerojosh take a look at this PR if they have the bandwidth.

**Comment by jeffdaily** [comment]
Rebased to pick up NCCL broadcast change.  Also updated the DDL sources, though I did not test them.

File: horovod/common/ops/gpu_operations.h [hunk 1]
```diff
@@ -14,70 +14,91 @@
...
 namespace common {
 
-struct CUDAContext {
-  cudaError_t GetCudaEvent(cudaEvent_t* event);
-
-  cudaError_t ReleaseCudaEvent(cudaEvent_t event);
+class GPUContext {
+public:
+  GPUContext();
+  ~GPUContext();
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply horovod/common/ops/gpu_operations.h]: I ran into an issue desribed here https://herbsutter.com/gotw/_100/ .  Specifically: 

> You still need to write the visible class’ destructor yourself and define it out of line in the implementation file, even if normally it’s the same as what the compiler would generate. This is because although both unique_ptr and shared_ptr can be instantiated with an incomplete type, unique_ptr’s destructor requires a complete type in order to invoke delete (unlike shared_ptr which captures more information when it’s constructed). By writing it yourself in the implementation file, you force it to be defined in a place where impl is already defined, and this successfully prevents the compiler from trying to automatically generate the destructor on demand in the caller’s code where impl is not defined.

[the remark this answers is not in the document]
Comment (nvcastet) [reply horovod/common/ops/gpu_operations.h]: Thanks @jeffdaily for the explanation and the link.

**Comment by jeffdaily** [comment]
@tgaddair are the last two test failures transient or do they indicate a real issue?  Also, this PR currently does not integrate any kind of CI build for ROCm yet.  What is your acceptance criteria?

**Comment by tgaddair** [comment]
@jeffdaily one of the errors appears to be real:
https://buildkite.com/horovod/horovod/builds/1844#c24a6237-3f92-4cc5-9310-bcf714008444

The MXNet test for mixed install is consistently failing because the Horovod MXNet plugin did not compile successfully.

Once we figure that out, I think we should be safe to land.

**Comment by jeffdaily** [comment]
@tgaddair  I tried building the docker container for the mixed install so I could see how horovod is failing to build the mxnet extension, and I'm seeing the following error:

```
  CMake Error: The source directory "/tmp/pip-req-build-7pquo6gf/third_party/gloo" does not exist.
  Specify --help for usage, or press the help button on the CMake GUI.
  INFO: Unable to build MXNet plugin, will skip it.
...
  error: None of TensorFlow, PyTorch, or MXNet plugins were built. See errors above.
  Building wheel for horovod (setup.py): finished with status 'error'
  ERROR: Failed building wheel for horovod
```

**Comment by jeffdaily** [comment]
@tgaddair disregard previous comment concerning the dockerfile build.  I forgot to fetch submodules prior to attempting the dockerfile build.

I was able to reproduce the mxnet build failure.  Fix will be ready shortly.

**Comment by jeffdaily** [comment]
@tgaddair fixed the MXNet test for mixed install.  Other errors have popped up now that were passing in the previous attempt.

**Comment by tgaddair** [comment]
Hey @jeffdaily, looks like test failures may be transient.  Rerunning them now.

File: horovod/common/ops/adasum_gpu_operations.cc [hunk 2]
```diff
@@ -172,33 +172,30 @@ AdasumCudaAllreduceOp::NcclHierarchical(std::vector<TensorTableEntry>& entries,
...
-      cuda_context_->RecordEvent(cuda_op_context_.event_queue, NCCL_REDUCE,
-                                 *cuda_op_context_.stream);
+      gpu_context_->RecordEvent(gpu_op_context_.event_queue, NCCL_REDUCE,
+                                 *gpu_op_context_.stream);
     }
   }
 
   if (global_state_->controller->IsHomogeneous() || is_root_rank) {
-    // cudaHostAlloc is significantly slower than malloc.  Pre-allocating
+    // gpuHostAlloc is significantly slower than malloc.  Pre-allocating
```
Comment (tgaddair) [inline horovod/common/ops/adasum_gpu_operations.cc]: For consistency with other comments and NVIDIA documentation link below, this should be left as `cudaHostAlloc`.

File: horovod/common/ops/nccl_operations.cc [hunk 3]
```diff
@@ -279,34 +279,32 @@ NCCLHierarchicalAllreduce::Execute(std::vector<TensorTableEntry>& entries,
...
     nccl_context_->ErrorCheck("ncclReduce", nccl_result);
     if (global_state_->timeline.Initialized()) {
-      cuda_context_->RecordEvent(cuda_op_context_.event_queue, NCCL_REDUCE, *cuda_op_context_.stream);
+      gpu_context_->RecordEvent(gpu_op_context_.event_queue, NCCL_REDUCE, *gpu_op_context_.stream);
     }
   }
 
   if (global_state_->controller->IsHomogeneous() || is_root_rank) {
-    // cudaHostAlloc is significantly slower than malloc.  Pre-allocating
+    // gpuHostAlloc is significantly slower than malloc.  Pre-allocating
```
Comment (tgaddair) [inline horovod/common/ops/nccl_operations.cc]: Same above: `gpuHostAlloc` -> `cudaHostAlloc`.

**Comment by tgaddair** [comment]
Hey @jeffdaily, took a closer look over the PR and discussed it with some other Horovod contributors.  Overall, I think the abstraction layer is good.  

The one thing I think we should still address is the possibility for the ROCm/RCCL and CUDA/NCCL APIs to diverge in the future.  With the pimpl pattern, we're locked in to the API exposed by the GPUContext.  In this design how would a contributor add functionality specific to one of these frameworks but not the other?

File: horovod/common/ops/adasum_gpu_operations.cc (hunk 2 above)
Comment (jeffdaily) [reply horovod/common/ops/adasum_gpu_operations.cc]: Done.

File: horovod/common/ops/nccl_operations.cc (hunk 3 above)
Comment (jeffdaily) [reply horovod/common/ops/nccl_operations.cc]: Done.

**Comment by jeffdaily** [comment]
I am rather at a loss to suggest how to address a future ROCm/RCCL and CUDA/NCCL API divergence.  With respect to HIP (ROCm) versus CUDA APIs, parity is intended by design.  I am not privy to the CUDA development roadmap, so I can't predict new or deprecated APIs.  Are there any plans for Horovod to use less common CUDA APIs?

Concerning RCCL versus NCCL, again parity is by design in order to make it easier to support AMD hardware through the same interface.

I think if there ever exists some CUDA or NCCL feature that isn't reflected by HIP or RCCL, we have the option of separating implementation details within the same source file using `#if HAVE_ROCM`, or a source file that is compiled for CUDA/NCCL only.  It would need to be documented as Horovod feature X is not supported for ROCm builds.

GPUDirect, or ROCmRDMA, is very well supported with AMD hardware.

**Comment by jeffdaily** [comment]
Rebased again to resolve conflicts from https://github.com/horovod/horovod/commit/d4d54bc7e772c1511a2634d6c0d7c6fc2456599f

File: setup.py [hunk 4]
```diff
@@ -608,17 +649,21 @@ def get_common_options(build_ext):
...
+    have_cuda = False
+    have_rocm = False
+    gpu_include_dirs = gpu_lib_dirs = gpu_macros = []
     if gpu_allreduce or gpu_allgather or gpu_broadcast:
-        have_cuda = True
-        cuda_include_dirs, cuda_lib_dirs = get_cuda_dirs(build_ext, cpp_flags)
-    else:
-        have_cuda = False
-        cuda_include_dirs = cuda_lib_dirs = []
+        if os.environ.get('HOROVOD_GPU_ROCM'):
```
Comment (tgaddair) [inline setup.py]: Can we get rid of this environment variable and ask the user to set `HOROVOD_GPU_ALLREDUCE=RCCL` instead?

**Comment by tgaddair** [comment]
@jeffdaily I'm fine punting on the API divergence thing for now.  I think the right longterm solution is to create separate NCCL vs RCCL derived classes with separate CUDA and ROCM implementations of the `GPUContext`.  But I think perhaps we can cross that bridge when we come to it, or doing it in a followup PR.

**Comment by nvcastet** [comment]
@jeffdaily @tgaddair Using derived classes would also cause issues because even if we use the base type through the code, we will still need to construct the object with the derived type and to avoid compilation errors, we would need have to have a bunch of "ifdef" which is not really clean code in modern C++. Were you thinking of another way to do that?
OOP polymorphism seems to work better when everything can get built and also in this case, both APIs will not be used in the same build.
The C++11 version of the pimpl pattern seems to be a good fit.
With any interface/encapsulation design, we have will have the potentiality of divergence in implementations. If one implementation does not support a feature, we can always document it or throw a not_implemented exception printing a clear error message.
Is anyone thinking of a better pattern? @jeffdaily is pimpl the main design pattern AMD has been using in open source projects where a single GPU API existed? I would assume so for C++ code.

File: setup.py (hunk 4 above)
Comment (jeffdaily) [reply setup.py]: I think at best we could use `HOROVOD_GPU_ALLREDUCE=RCCL` to automatically set a missing `HOROVOD_GPU_ROCM`, but shouldn't we still support builds where RCCL is not used but we still build horovod for ROCm and assume a GPUDirect-capable MPI install?  Is this still a use case for CUDA?

Comment (nvcastet) [reply setup.py]: Correct, now we can use cuda-aware MPI without NCCL if we say (HOROVOD_GPU_ALLREDUCE=MPI or HOROVOD_GPU_BROADCAST=MPI).

**Comment by jeffdaily** [comment]
@nvcastet unfortunately, there is not one main design pattern for supporting AMD hardware in open source projects.

As one might expect, long-established open source projects tend to resist adding HIP APIs, and certainly not fully converting code bases to use HIP alone even though HIP can compile to CUDA.  That said, I've been on the teams to support TensorFlow, PyTorch, and now Horovod -- and each project maintainer has dictated a different approach.

TensorFlow already had a device abstraction layer to differentiate between CPU and GPU devices.  The GPU was itself further abstracted as a stream executor.  That was all done prior to our involvement, but it made the most sense to integrate our ROCm stack as an implementation of the stream executor abstract interface.  Even so, there are plenty of cases where APIs or device features are not the same.  For example, CUDNN has either different behavior or a different API interface compared to our MIOpen, and so such code must be protected with `#if TENSORFLOW_USE_ROCM`.  In summary, TensorFlow uses both this preprocessor macro as well as separately compiled sources even though attempts were made to use abstract interfaces for the GPU device and libraries (BLAS, CUDNN, etc).

PyTorch has mandated that the code base is "hipified" during the build, so a"hipify" python script was developed and maintained as part of the PyTorch code base.  There are very few pure-HIP source files; the rest of the PyTorch source looks like any other CUDA-based project.  But, macros can still be used to project CUDA-only code and features.  We have further been mandated to use this hipify script for building all ROCm PyTorch extensions.  Those changes are prepared in a separate branch that will become a new PR if this current PR is accepted.

File: setup.py (hunk 4 above)
Comment (tgaddair) [reply setup.py]: That's a good point.  How about we change it to `HOROVOD_GPU=ROCM`.  That way we leave open the possibility of other APIs as well, and potentially to auto-infer if not set.

**Comment by tgaddair** [comment]
@nvcastet Currently we use preprocessor macros to determine which set of ops to use [here](https://github.com/horovod/horovod/blob/master/horovod/common/operations.cc#L142).  Otherwise, the code is isolated into its own classes / files by framework so it's easier for developers to reason about the code paths.

I think that if the APIs ever diverged, we would necessarily need to either do something similar, or mix the preprocessor macros into the `nccl_operations.cc` file, the latter of which I think is less ideal from a maintainability standpoint.

So long as the APIs are identical, I agree the pimpl pattern is cleaner.  And since there isn't any indication that they will diverge at the moment, I also agree it's the better solution for now.

File: setup.py (hunk 4 above)
Comment (jeffdaily) [reply setup.py]: I'm assuming you would like the current behavior to remain the same, i.e., assume that CUDA is selected.  So something like the following?

```python
    have_cuda = False
    have_rocm = False
    gpu_include_dirs = gpu_lib_dirs = gpu_macros = []
    if gpu_allreduce or gpu_allgather or gpu_broadcast:
        gpu_type = os.environ.get('HOROVOD_GPU', 'CUDA')
        if gpu_type == 'CUDA':
            have_cuda = True
            gpu_include_dirs, gpu_lib_dirs = get_cuda_dirs(build_ext, cpp_flags)
        elif gpu_type == 'ROCM':
            have_rocm = True
            gpu_include_dirs, gpu_lib_dirs, gpu_macros = get_rocm_dirs(build_ext, cpp_flags)
        else:
            raise DistutilsError("Unknown HOROVOD_GPU type '%s'" % gpu_type)
```

Comment (tgaddair) [reply setup.py]: Yes, that's exactly what I had in mind.  Does that sound good to you?

Comment (jeffdaily) [reply setup.py]: Done.
